fili-core/src/main/java/com/yahoo/bard/webservice/application/AbstractBinderFactory.java

```diff
@@ -919,7 +924,24 @@ protected JobPayloadBuilder buildJobPayloadBuilder() {
      * @return A DruidWebService
      */
     protected DruidWebService buildDruidWebService(DruidServiceConfig druidServiceConfig, ObjectMapper mapper) {
-        return new AsyncDruidWebServiceImpl(druidServiceConfig, mapper, HashMap::new);
+        Supplier<Map<String, String>> supplier = HashMap::new;
+        String customSupplierClassString = SYSTEM_CONFIG.getStringProperty(DRUID_HEADER_SUPPLIER_CLASS, null);
+        if (customSupplierClassString != null && customSupplierClassString != "") {
+            try {
+                Class<?> c = Class.forName(customSupplierClassString);
+                Constructor<?> constructor = c.getConstructor();
+                supplier = (Supplier<Map<String, String>>) constructor.newInstance();
+            } catch (ClassNotFoundException | NoSuchMethodException |
+                    IllegalAccessException | InstantiationException | InvocationTargetException e) {
+                LOG.error(
+                        "Unable to load the Druid query header supplier, className: {}, exception: {}",
+                        customSupplierClassString,
+                        e
+                );
+                throw new IllegalStateException(e);
+            }
+        }
```

Comment:
I think we should pull the `buildDruidWebServiceHeaderSupplier` code into it's own method. That will give another (fairly easy) override-based config mechanism if someone's `Supplier` needs more than a no-param constructor.